src/services/blockchain/clients/stellar/client.rs

```diff
@@ -45,6 +54,64 @@ impl<T: Send + Sync + Clone> StellarClient<T> {
 	pub fn new_with_transport(http_client: T) -> Self {
 		Self { http_client }
 	}
+
+	/// Checks a successfully received JSON RPC response body for an "error" field.
+	/// If an error field is present, it attempts to identify specific conditions like
+	/// "outside of retention window" or handles it as a generic RPC error.
+	fn check_and_handle_rpc_error(
+		&self,
+		response_body: &serde_json::Value,
+		start_sequence: u32,
+		target_sequence: u32,
+		method_name: &'static str,
+	) -> Result<(), StellarClientError> {
+		if let Some(json_rpc_error) = response_body.get("error") {
+			let rpc_code = json_rpc_error
+				.get("code")
+				.and_then(|c| c.as_i64())
+				.unwrap_or(0);
+			let rpc_message = json_rpc_error
+				.get("message")
+				.and_then(|m| m.as_str())
+				.unwrap_or("Unknown RPC error")
+				.to_string();
+
+			// Specifically check for out-of-retention (code -32600 for Soroban RPC)
+			// There are two possible error messages that indicate this condition:
+			// - get transactions: "startLedger must be within the ledger range: 57317310 - 57369149"
+			// - get events: "start ledger must be between the oldest ledger: 57317319 and the latest ledger: 57369158 for this rpc instance"
+			if rpc_code == -32600
+				&& (rpc_message
+					.to_lowercase()
+					.contains("must be within the ledger range")
+					|| rpc_message
+						.to_lowercase()
+						.contains("must be between the oldest ledger"))
+			{
```

Comment:
nit; We could have these error strings in an array that we check against instead. I expect they may change later once Stellar normalises their RPC responses, so it would be more straightforward to add/remove error strings then.

Reply:
FYI, looks like there's already a fix merged, just not released yet: https://github.com/stellar/stellar-rpc/issues/454#issuecomment-2960229227